README.md
# &lt;visual-logger&gt;

[![Build Status](https://travis-ci.com/kcmr/visual-logger.svg?branch=master)](https://travis-ci.com/kcmr/visual-logger)
[![npm version](https://badge.fury.io/js/%40kuscamara%2Fvisual-logger.svg)](https://badge.fury.io/js/%40kuscamara%2Fvisual-logger)
[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/@kuscamara/visual-logger)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![codecov](https://codecov.io/gh/kcmr/visual-logger/branch/master/graph/badge.svg)](https://codecov.io/gh/kcmr/visual-logger)
![Dependency status](https://img.shields.io/david/kcmr/visual-logger.svg)


> Displays calls to `window.console` methods in a visual terminal using [Xterm.js](https://xtermjs.org/).

[![Visual Logger demo](https://github.com/kcmr/visual-logger/raw/master/visual-logger.png)](https://visual-logger.now.sh/)

🚀 **Demo:** [https://visual-logger.now.sh/](https://visual-logger.now.sh/)

This webcomponent follows the [open-wc](https://github.com/open-wc/open-wc) recommendation.

## Installation

```bash
npm i @kuscamara/visual-logger
```

## Usage

⚠️ **Important:** Due to differences in module bundlers ([see issue in xtermjs repo](https://github.com/xtermjs/xterm.js/issues/2486)) handling exports, the component does not import `xterm` by itself and uses `window.Terminal` `constructor`, so **`xterm` should be loaded in `window` before the component**.   
`xterm` is also required as `devDependency` for the tests.

Include the component in your page and start using `window.console` methods.

```html
<script src="https://unpkg.com/xterm@4.4.0/lib/xterm.js"></script>
<script type="module" src="node_modules/@kuscamara/visual-logger/visual-logger.js"></script>

<visual-logger></visual-logger>
```

## Options

### [API](api.md)

Check out the **[API docs](api.md)** for the complete list of properties / attributes and methods.

### Disabling browser logging (`window.console`)

Set `noConsole` to `true`.

```html
<visual-logger no-console></visual-logger>
```

### Excluding `console` methods

Config the excluded methods in `excludedLogMethods` (`array`)

```html
<visual-logger excluded-log-methods='["error"]'></visual-logger>
```

### Xterm styles

The required styles for the terminal are loaded by default from a CDN (cdnjs.cloudflare.com), so you don't need to import them manually. If you need to load the stylesheet from a different location, use the `stylesheet-uri` attribute to specify the path.

```html
<visual-logger stylesheet-uri="node_modules/xterm/css/xterm.css"></visual-logger>
```

### Changing size

The height (`rows`) and column width (`cols`) can be changed after or before initialization.

```html
<visual-logger rows="10" cols="120"></visual-logger>
```

### Customizing styles

The default terminal styles (background, ANSI colors and font styles) can be customized using the [`theme` option of Xterm.js](https://xtermjs.org/docs/api/terminal/interfaces/itheme/).

The component **does not use Shadow DOM**, so it can be customized from the outside using `visual-editor` tag.

```css
visual-editor > div {
  padding: 20px;
}
```

### Programmatically usage

Each of the `window.console` methods (`log`, `warn`, `info`, `error`) has a corresponding component method that accepts the same params.

_Note: only the first two params are shown in the terminal._

```html
<visual-logger></visual-logger>

<script>
  document.querySelector('visual-logger').log('Hello world!');
</script>
```

## Development

The following commands are available for development:

- `npm start`: Starts the development server.
- `npm t`: Runs the tests with coverage output.
- `npm run test:watch`: Runs the test in watch mode. The browser runner is available at http://localhost:9876/
- `npm run lint`: Runs [web component analyzer](https://www.npmjs.com/package/web-component-analyzer), eslint and prettier.
- `npm run format`: Runs linters fixing errors.
- `npm run docs`: Updates the API docs in [api.md](api.md) file and `custom-elements.json`.

## License

This project is licensed under the [MIT License](LICENSE).


CHANGELOG.md
# 1.0.0 (2020-03-07)


### Bug Fixes

* **package:** build script ([cfa4720](https://github.com/kcmr/visual-logger/commit/cfa47206db44a4c471bd32577e814f870587d9e0))


### Features

* first release ([cf3ac7a](https://github.com/kcmr/visual-logger/commit/cf3ac7a5a2b9c41c3aec523c7811bc55762a583f))


test/visual-logger.test.js
import { html, fixture, fixtureCleanup, assert } from '@open-wc/testing';
import sinon from 'sinon';
import 'xterm';
import { STYLESHEET_URI as DEFAULT_STYLESHEET_URI, ANSI_COLOR } from '../src/VisualLogger.js';
import '../visual-logger.js';

suite('<visual-logger>', () => {
  teardown(() => {
    fixtureCleanup();
  });

  suite('Settings', () => {
    test('renders a terminal with the settings configured in "cols", "rows", "lineHeight"', async () => {
      const settings = {
        cols: 2,
        rows: 10,
        lineHeight: 2,
      };

      const terminal = new window.Terminal();
      const terminalProto = Object.getPrototypeOf(terminal);
      const terminalOpenStub = sinon.stub(terminalProto, 'open');

      const el = await fixture(html`
        <visual-logger
          cols="${settings.cols}"
          rows="${settings.rows}"
          line-height="${settings.lineHeight}"
        ></visual-logger>
      `);

      await el.updateComplete;

      assert.isTrue(terminalOpenStub.calledWith(el), 'opens a terminal in its root');
      assert.instanceOf(el.terminal, terminalProto.constructor);

      Object.entries(settings).forEach(([setting, value]) => {
        assert.strictEqual(el.terminal.getOption(setting), value, `${setting} is ${value}`);
      });
    });

    test('injects terminal styles from CDN if "stylesheetUri" is not set', async () => {
      await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const terminalStyles = document.querySelector('#xterm-styles');
      assert.strictEqual(terminalStyles.getAttribute('href'), DEFAULT_STYLESHEET_URI);
    });

    test('changing "cols" after initialization resizes the terminal', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const resizeSpy = sinon.spy(el.terminal, 'resize');

      el.cols = 3;
      await el.updateComplete;

      const colsParam = resizeSpy.getCall(0).args[0];

      assert.strictEqual(colsParam, 3);
    });

    test('changing "rows" after initialization resizes the terminal', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const resizeSpy = sinon.spy(el.terminal, 'resize');

      el.rows = 3;
      await el.updateComplete;

      const rowsParam = resizeSpy.getCall(0).args[1];

      assert.strictEqual(rowsParam, 3);
    });

    test('changing "lineHeight" after initialization updates the terminal "lineHeight"', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const setOptionSpy = sinon.spy(el.terminal, 'setOption');

      el.lineHeight = 2;
      await el.updateComplete;

      assert.isTrue(setOptionSpy.calledWith('lineHeight', 2));
    });
  });

  suite('Methods', () => {
    test('"log()" logs a green message', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      el.log('foo');

      const message = `${ANSI_COLOR.green}foo`;

      assert.isTrue(writelnSpy.calledWith(message));
    });

    test('"error()" logs a red message', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      el.error('foo');

      const message = `${ANSI_COLOR.red}foo`;

      assert.isTrue(writelnSpy.calledWith(message));
    });

    test('"warn()" logs a yellow message', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      el.warn('foo');

      const message = `${ANSI_COLOR.yellow}foo`;

      assert.isTrue(writelnSpy.calledWith(message));
    });

    test('"info()" logs a blue message', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      el.info('foo');

      const message = `${ANSI_COLOR.blue}foo`;

      assert.isTrue(writelnSpy.calledWith(message));
    });

    test('"clear() clears the terminal', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );
      const resetSpy = sinon.spy(el.terminal, 'reset');

      el.clear();

      assert.isTrue(resetSpy.called);
    });
  });

  suite('Logging', () => {
    test('calling "console.log()" calls "log()" with the same params', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      // cannot spy on log method
      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      console.log('first', 'second');
      const writelnSpyParam = writelnSpy.getCall(0).args[0];

      assert.include(writelnSpyParam, ANSI_COLOR.green);
      assert.include(writelnSpyParam, 'first');
      assert.include(writelnSpyParam, 'second');
    });

    test('calling "console.info()" calls "info()" with the same params', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      console.info('first', 'second');
      const writelnSpyParam = writelnSpy.getCall(0).args[0];

      assert.include(writelnSpyParam, ANSI_COLOR.blue);
      assert.include(writelnSpyParam, 'first');
      assert.include(writelnSpyParam, 'second');
    });

    test('calling "console.warn()" calls "warn()" with the same params', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      console.warn('first', 'second');
      const writelnSpyParam = writelnSpy.getCall(0).args[0];

      assert.include(writelnSpyParam, ANSI_COLOR.yellow);
      assert.include(writelnSpyParam, 'first');
      assert.include(writelnSpyParam, 'second');
    });

    test('calling "console.error()" calls "error()" with the same params', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');

      console.error('first', 'second');
      const writelnSpyParam = writelnSpy.getCall(0).args[0];

      assert.include(writelnSpyParam, ANSI_COLOR.red);
      assert.include(writelnSpyParam, 'first');
      assert.include(writelnSpyParam, 'second');
    });

    test('calling a console method included in "excludedLogMethods" does not call the corresponding component method', async () => {
      const el = await fixture(
        // eslint-disable-next-line lit/attribute-value-entities
        html`
          <visual-logger excluded-log-methods='["log"]'></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      console.log('not logged');

      assert.isFalse(writelnSpy.called);
    });

    test('shows a character separator (‣) between the first and second log param', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      console.log('first', 'second');
      const expectedMessage = `${ANSI_COLOR.green}first ‣ second`;

      assert.isTrue(writelnSpy.calledWith(expectedMessage));
    });

    test('param separator is not shown if console is called when one param', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      console.log('first');
      const expectedMessage = `${ANSI_COLOR.green}first`;

      assert.isTrue(writelnSpy.calledWith(expectedMessage));
    });

    test('console[method] second param is shown if it is 0', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      console.log('first', 0);
      const expectedMessage = `${ANSI_COLOR.green}first ‣ 0`;

      assert.isTrue(writelnSpy.calledWith(expectedMessage));
    });

    test('objects in second param are stringified', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      const obj = { foo: 'bar' };
      console.log('first', obj);
      const expectedMessage = `${ANSI_COLOR.green}first ‣ {"foo":"bar"}`;

      assert.isTrue(writelnSpy.calledWith(expectedMessage));
    });

    test('arrays in second param are stringified', async () => {
      const el = await fixture(
        html`
          <visual-logger></visual-logger>
        `,
      );

      const writelnSpy = sinon.spy(el.terminal, 'writeln');
      const arr = ['uno', 'dos'];
      console.log('first', arr);
      const expectedMessage = `${ANSI_COLOR.green}first ‣ ["uno","dos"]`;

      assert.isTrue(writelnSpy.calledWith(expectedMessage));
    });
  });
});


test/browser-console.test.js
import { html, fixture, expect } from '@open-wc/testing';

import '../browser-console.js';

describe('BrowserConsole', () => {
  it('has a default title "Hey there" and counter 5', async () => {
    const el = await fixture(html`
      <browser-console></browser-console>
    `);

    expect(el.title).to.equal('Hey there');
    expect(el.counter).to.equal(5);
  });

  it('increases the counter on button click', async () => {
    const el = await fixture(html`
      <browser-console></browser-console>
    `);
    el.shadowRoot.querySelector('button').click();

    expect(el.counter).to.equal(6);
  });

  it('can override the title via attribute', async () => {
    const el = await fixture(html`
      <browser-console title="attribute title"></browser-console>
    `);

    expect(el.title).to.equal('attribute title');
  });

  it('passes the a11y audit', async () => {
    const el = await fixture(html`
      <browser-console></browser-console>
    `);

    await expect(el).shadowDom.to.be.accessible();
  });
});


karma.conf.js
/* eslint-disable import/no-extraneous-dependencies */
const { createDefaultConfig } = require('@open-wc/testing-karma');
const merge = require('deepmerge');

module.exports = config => {
  config.set(
    merge(createDefaultConfig(config), {
      files: [
        // runs all files ending with .test in the test folder,
        // can be overwritten by passing a --grep flag. examples:
        //
        // npm run test -- --grep test/foo/bar.test.js
        // npm run test -- --grep test/bar/*
        { pattern: config.grep ? config.grep : 'test/**/*.test.js', type: 'module' },
      ],

      client: {
        mocha: {
          ui: 'tdd',
        },
      },

      browserConsoleLogOptions: {
        level: config.logLevel,
      },

      coverageIstanbulReporter: {
        thresholds: {
          global: {
            statements: 98,
            branches: 88,
            functions: 100,
            lines: 98,
          },
        },
      },

      esm: {
        nodeResolve: true,
      },
    }),
  );

  return config;
};


src/BrowserConsole.js
/* eslint-disable no-console */
import { html, LitElement } from 'lit-element';
import 'xterm';

const STYLESHEET_URI = 'https://cdnjs.cloudflare.com/ajax/libs/xterm/3.14.5/xterm.min.css';
const ANSI_COLOR = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[36m',
};

function format(param = '') {
  if (typeof param === 'object') {
    try {
      return JSON.stringify(param);
    } catch (error) {
      console.error(error);
    }
  }

  return param;
}

function getMessage(...args) {
  const [first, second] = args;
  return [first, format(second)].filter(Boolean).join(' ‣ ');
}

export class BrowserConsole extends LitElement {
  static get properties() {
    return {
      /**
       * Console columns
       */
      cols: {
        type: Number,
      },

      /**
       * Console rows
       */
      rows: {
        type: Number,
      },

      /**
       * Line height
       */
      lineHeight: {
        type: Number,
        attribute: 'line-height',
      },

      /**
       * Terminal theme
       * @see https://xtermjs.org/docs/api/terminal/interfaces/itheme/
       */
      theme: {
        type: Object,
        attribute: false,
      },

      /**
       * Enabled log methods (window.console)
       */
      logMethods: {
        type: Array,
        attribute: 'log-methods',
      },

      /**
       * Supress window.console logs
       */
      noConsole: {
        type: Boolean,
        attribute: 'no-console',
      },

      /**
       * URI of the Terminal stylesheet (Uses cdnjs if not provided)
       */
      stylesheetUri: {
        type: String,
        attribute: 'stylesheet-uri',
      },
    };
  }

  constructor() {
    super();
    this.cols = 200;
    this.rows = 5;
    this.lineHeight = 1.5;
    this.logMethods = ['log', 'info', 'warn', 'error'];
    this.stylesheetUri = STYLESHEET_URI;
  }

  createRenderRoot() {
    return this;
  }

  firstUpdated() {
    this._loadStyles();
    this._initialize();
  }

  _loadStyles() {
    const style = document.createElement('link');
    document.head.appendChild(
      Object.assign(style, {
        rel: 'stylesheet',
        href: this.stylesheetUri,
      }),
    );
  }

  _initialize() {
    this._terminal = new window.Terminal({
      cols: this.cols,
      rows: this.rows,
      lineHeight: this.lineHeight,
      theme: this.theme,
    });

    this._terminal.open(this);
    this._spyConsole();
  }

  _spyConsole() {
    const consoleEnabled = !this.noConsole;

    function logDecorator(fn, method) {
      return function log(...args) {
        method(...args);

        if (consoleEnabled) {
          fn.call(this, ...args);
        }
      };
    }

    this.logMethods.forEach(method => {
      const logMethod = this[method].bind(this);
      console[method] = logDecorator(console[method], logMethod);
    });
  }

  /**
   * Logs a message (console.log)
   */
  log(...args) {
    this._writeln('green', getMessage(...args));
  }

  /**
   * Logs an error (console.error)
   */
  error(...args) {
    this._writeln('red', getMessage(...args));
  }

  /**
   * Logs a warning (console.warn)
   */
  warn(...args) {
    this._writeln('yellow', getMessage(...args));
  }

  /**
   * Logs an info (console.info)
   */
  info(...args) {
    this._writeln('blue', getMessage(...args));
  }

  /**
   * Clears the console
   */
  clear() {
    this._terminal.reset();
  }

  _writeln(color, message) {
    this._terminal.writeln(`${ANSI_COLOR[color]}${message}`);
  }

  render() {
    return html`
      <style>
        browser-console {
          display: block;
          position: fixed;
          bottom: 0;
          width: 100%;
          box-sizing: border-box;
        }

        browser-console > div {
          padding: 10px;
        }

        browser-console code {
          font-family: monospace, monospace;
        }
      </style>
    `;
  }
}


src/VisualLogger.js
/* eslint-disable no-console */
import { html, LitElement } from 'lit-element';

const LOG_METHODS = ['log', 'warn', 'info', 'error'];
export const STYLESHEET_URI = 'https://cdnjs.cloudflare.com/ajax/libs/xterm/3.14.5/xterm.min.css';
export const ANSI_COLOR = {
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[36m',
};

function format(param = '') {
  if (typeof param === 'object') {
    try {
      return JSON.stringify(param);
    } catch (error) {
      console.error(error);
    }
  }

  return param;
}

function getMessage(...args) {
  const [first, second] = args;
  const notEmpty = value => Boolean(String(value));

  return [first, format(second)].filter(notEmpty).join(' ‣ ');
}

/**
 * Browser UI terminal using Xterm.js to display console methods (log, warn, error, info).
 * @element visual-logger
 */
export class VisualLogger extends LitElement {
  static get properties() {
    return {
      /**
       * Terminal columns
       * @attr cols
       */
      cols: {
        type: Number,
      },

      /**
       * Terminal rows
       * @attr rows
       */
      rows: {
        type: Number,
      },

      /**
       * Line height
       * @attr line-height
       */
      lineHeight: {
        type: Number,
        attribute: 'line-height',
      },

      /**
       * Terminal theme
       * @see https://xtermjs.org/docs/api/terminal/interfaces/itheme/
       */
      theme: {
        type: Object,
        attribute: false,
      },

      /**
       * Excluded log (window.console) methods
       * @attr excluded-log-methods
       */
      excludedLogMethods: {
        type: Array,
        attribute: 'excluded-log-methods',
      },

      /**
       * Supress window.console logs
       * @attr no-console
       */
      noConsole: {
        type: Boolean,
        attribute: 'no-console',
      },

      /**
       * URI of the Terminal stylesheet (Uses cdnjs if not provided)
       * @attr stylesheet-uri
       */
      stylesheetUri: {
        type: String,
        attribute: 'stylesheet-uri',
      },
    };
  }

  /**
   * Returns the xterm instance
   */
  get terminal() {
    return this._terminal;
  }

  constructor() {
    super();
    this.cols = 200;
    this.rows = 5;
    this.lineHeight = 1.5;
    this.excludedLogMethods = [];
    this.stylesheetUri = STYLESHEET_URI;
  }

  /**
   * @private
   */
  createRenderRoot() {
    return this;
  }

  firstUpdated() {
    this._loadStyles();
    this._initialize();
  }

  updated(changedProperties) {
    super.updated(changedProperties);
    this._updateTerminal(changedProperties);
  }

  _loadStyles() {
    const style = document.createElement('link');
    document.head.appendChild(
      Object.assign(style, {
        id: 'xterm-styles',
        rel: 'stylesheet',
        href: this.stylesheetUri,
      }),
    );
  }

  _initialize() {
    this._terminal = new window.Terminal({
      cols: this.cols,
      rows: this.rows,
      lineHeight: this.lineHeight,
      theme: this.theme,
    });

    this._terminal.open(this);
    this._spyConsole();
  }

  _updateTerminal(changedProperties) {
    if (
      (changedProperties.has('cols') && this.cols !== undefined) ||
      (changedProperties.has('rows') && this.rows !== undefined)
    ) {
      this._terminal.resize(this.cols, this.rows);
    }

    if (changedProperties.has('lineHeight') && this.lineHeight !== undefined) {
      this._terminal.setOption('lineHeight', this.lineHeight);
    }
  }

  _spyConsole() {
    const consoleEnabled = !this.noConsole;
    const allowedLogMethods = LOG_METHODS.filter(
      method => !this.excludedLogMethods.includes(method),
    );

    function logDecorator(fn, method) {
      return function log(...args) {
        method(...args);

        if (consoleEnabled) {
          fn.call(this, ...args);
        }
      };
    }

    allowedLogMethods.forEach(method => {
      const logMethod = this[method].bind(this);
      console[method] = logDecorator(console[method], logMethod);
    });
  }

  /**
   * Logs a message (console.log)
   */
  log(...args) {
    this._writeln('green', getMessage(...args));
  }

  /**
   * Logs an error (console.error)
   */
  error(...args) {
    this._writeln('red', getMessage(...args));
  }

  /**
   * Logs a warning (console.warn)
   */
  warn(...args) {
    this._writeln('yellow', getMessage(...args));
  }

  /**
   * Logs an info (console.info)
   */
  info(...args) {
    this._writeln('blue', getMessage(...args));
  }

  /**
   * Clears the console
   */
  clear() {
    this._terminal.reset();
  }

  _writeln(color, message) {
    this._terminal.writeln(`${ANSI_COLOR[color]}${message}`);
  }

  render() {
    return html`
      <style>
        visual-logger {
          display: block;
          position: fixed;
          bottom: 0;
          left: 0;
          width: 100%;
        }

        visual-logger > div {
          padding: 10px;
        }
      </style>
    `;
  }
}
